src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -1480,6 +1481,7 @@ internal override bool IsMetadataOptional
         }
 
         internal sealed override bool IsMetadataIn
+            // PROTOTYPE: or HasRequiresLocationAttribute
```

Comment:
Not related to this PR: Are we going to allow the `RequiresLocationAttribute` to be specified in source? 
We often disallow compiler-specific attributes in source (see `ERR_ExplicitParamArray` for example), although I'm not sure why the `In` and `Out` attributes were not handled that way (possibly due to legacy behavior?).

Reply:
Good catch. I guess it makes sense to disallow it.

Reply:
`[In]` and `[Out]` are special runtime attributes that are built directly into the metadata. They were primarily designed for interop purposes and the languages uses them to allow its own features to be tracked while also allowing correct semantics for the features in interop scenarios. The language does disallow them to be specified in source in some, but not all scenarios.

In particular, the language disallows `[Out] ref T x` and expects you to use `out T x` instead. The language likewise disallows `[In] out T x` and `[In, Out] out T x` because `[In]` on an `out` parameter is non-sensical and is non-distinguishable at the IL level since in IL it is encoded as `[Out] ref T x`. The language does not disallow `[Out] out int x`. It is redundant information, but allowed regardless.

The language does not disallow `[In] ref T x` because the language did not have `in T x` in v1.0. The language likewise annotated `in T x` in metadata using both `[In]` and `[IsReadOnly]` because the latter is what is used to identify the feature since it would have otherwise been a source breaking change due to `ref T x` being originally disallowed on `in` parameters. The language does not allow `[IsReadOnly]` to be manually specified in any scenario. The language does not allow `[Out] in T x` or `[In, Out] in T x` to be specified since `[Out]` on an `in` parameter is non-sensical.

The language does allow `[In, Out] ref T x` to be specified even though it is redundant information. However, since `ref T x` does not require additional metadata to be recognized, there is a metadata difference in what is emitted.

Outside of `byrefs` (that is `in`, `out`, `ref`), the `[In]` and `[Out]` parameters may also be used on arbitrary `T` or `T[]` as it can also impact the interop/marshaling semantics in those scenarios. The language has no restrictions (that I'm aware of) when used in this way.

Reply:
Given that, I would expect that `[Out] ref readonly T x` and `[In, Out] ref readonly T x` to be disallowed and to surface something similar to `error CS8355: An in parameter cannot have the Out attribute.`.

I would expect that `[In] ref readonly T x` to be allowed.

I would expect that `[RequiresLocation] ref T x` to be disallowed and users expected to use `ref readonly T x` instead.